status_quo: alan creates a hanging alarm (#1)
webern/wg-async-foundations: src/vision/status_quo/bottlerocket_canary_hangs.md

# Bottlerocket Canary Hangs

[Bottlerocket] is a Linux-based open-source operating system that is purpose-built by Amazon Web Services for running containers.
We use our own program [pubsys] to ensure that Bottlerocket update repositories are healthy.

## The TL;DR

- It is confusing and dangerous that multiple tokio runtimes can panic or hang at program runtime.
- It is challenging that using multiple major versions of tokio (which is allowed by cargo) can fail at runtime.
- It is unfortunate that we need a 3rd party runtime in order to `block_on` a future, even if we are not trying to write async code.

## Multiple Tokio Runtimes

We use our own [tough] library to read and write TUF repositories.
This library was created before async became widespread and [reqwest] changed its main interface to `async`.
When reqwest switched to async, we used the `reqwest::blocking` feature instead of re-writing tough to be an async interface.
(Maybe we [should](https://github.com/awslabs/tough/issues/213) make tough an async interface, but we haven't yet.)

> In order to provide a non-async runtime, reqwest creates a tokio runtime so that it can await the futures.

Author: ```suggestion
In order to provide a non-async runtime, reqwest creates a tokio runtime so that it can await futures.
```

> Using the above libraries in pubsys, we created some parallel downloading logic.

Author: ```suggestion
In pubsys we created some parallel downloading logic while using the above libraries.
```

Without realizing the danger, we created a tokio runtime in pubsys and used futures/await to do this parallelization.
Surprisingly, in retrospect, this worked... until it didn't.

> Recently we had a high severity incident when we realized that our repository verification alarm was hanging.

Author: ```suggestion
Recently we we realized that our repository verification alarm was hanging.
```

> The root cause was that multiple tokio runtimes caused the hang.

Author: ```suggestion
The root cause was having multiple tokio runtimes, though we don't know what change exposed the issue.
```

> The fix was to eliminate the need for a tokio runtime and do the parallel downloads in a different way

Author: ```suggestion
The fix was to eliminate the need for a tokio runtime in the pubsys code path by doing the parallel downloads in a different way
```
